rust/core-lib/src/layers.rs

```diff
@@ -59,38 +62,72 @@ impl Scopes {
+    /// as edits occur.
+    pub fn update_all(&mut self, iv: Interval, len: usize) {
+        self.merged.edit(iv, SpansBuilder::new(len).build());
+        let empty_spans = SpansBuilder::new(len).build();
         for layer in self.layers.values_mut() {
-            layer.update_spans(iv, spans.clone());
+            layer.update_spans(iv, &empty_spans);
         }
+        self.resolve_styles(iv);
     }
 
     /// Updates the scope spans for a given layer.
     pub fn update_layer(&mut self, layer: PluginPid, iv: Interval, spans: Spans<u32>) {
         self.create_if_missing(layer);
-        self.layers.get_mut(&layer).unwrap().update_spans(iv, spans);
+        self.layers.get_mut(&layer).unwrap().update_spans(iv, &spans);
+        self.resolve_styles(iv);
     }
 
     /// Removes a given layer. This will remove all styles derived from
     /// that layer's scopes.
     pub fn remove_layer(&mut self, layer: PluginPid) -> Option<ScopeLayer> {
-        self.layers.remove(&layer)
+        let layer = self.layers.remove(&layer);
+        if layer.is_some() {
+            let iv_all = Interval::new_closed_closed(0, self.merged.len());
+            //TODO: should Spans<T> have a clear() method?
+            self.merged = SpansBuilder::new(self.merged.len()).build();
+            self.resolve_styles(iv_all);
+        }
+        layer
+    }
+
+    pub fn theme_changed<W: Write>(&mut self, doc_ctx: &DocumentCtx<W>) {
+        for layer in self.layers.values_mut() {
+            layer.theme_changed(doc_ctx);
+        }
+        self.merged = SpansBuilder::new(self.merged.len()).build();
+        let iv_all = Interval::new_closed_closed(0, self.merged.len());
+        self.resolve_styles(iv_all);
     }
 
-    /// For a given Interval, generates styles from scopes, resolving conflicts.
-    pub fn resolve_styles(&self, iv: Interval) -> Spans<Style> {
+    /// Resolves styles from all layers for the given interval, updating
+    /// the master style spans.
+    fn resolve_styles(&mut self, iv: Interval) {
         if self.layers.is_empty() {
-            return SpansBuilder::new(iv.size()).build()
+            return
         }
-        //TODO: implement layer merging.
-        let mut sb = SpansBuilder::new(iv.size());
-        let layer = self.layers.values().next().unwrap();
-        for (iv, val) in layer.spans.subseq(iv).iter() {
-            let style = layer.style_lookup[*val as usize];
-            sb.add_span(iv, style);
+        let mut layer_iter = self.layers.values();
+        let mut resolved = layer_iter.next().unwrap().subseq_styles(iv);
+
+        for other in layer_iter {
+            //FIXME: this creates a whole lot of unnecessary styles, because
+            // subseq_styles creates a new spans object and a new Style for each
+            // id. Better would be to rewrite Spans::merge to be a bit more like
+            // a reduce/fold operation, and then we could just use &Styles.
```

Comment:
I'd prefer that you do this before this PR is merged, but it's not a strong preference and I'm okay with it being merged before then.